#### Repo URL:
https://github.com/<redacted>/assignment-4-equation-of-a-slime-<redacted>

#### Commit ID:
af86c92abbf9af070037369f7394a586acbd162c


In [16]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix


## Part 1. Loading the dataset

In [6]:
# Using pandas load the dataset (load remotely, not locally)
df = pd.read_csv('science_data_large.csv')

# Output the first 15 rows of the data
print(df.head(15))

# Display a summary of the table information (number of datapoints, etc.)
print(df.info)

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<bound method DataFrame.info of      Temperature °C  Mols KCL     Size nm^3
0               469       647  6.244743e+05
1               403       694  5.779610e+05
2               302       975  6.196847e+05
3               779       916  1.460449e+06
4               901        18  4.325726e+04
..              

## Part 2. Splitting the dataset

In [7]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = df.drop(columns=['Size nm^3'])
y = df['Size nm^3']

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.1, random_state = 42)

## Part 3. Perform a Linear Regression

In [12]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
sample_datapoint = pd.DataFrame({
    'Temperature °C': [X_train['Temperature °C'].mean()],
    'Mols KCL': [X_train['Mols KCL'].mean()]
})

predicted_value = model.predict(sample_datapoint)
print(f"Predicted output: {predicted_value[0]}")

# Report on the score for that model, in your own words (markdown, not code) explain what the score means
score = model.score(X_test, y_test)
print(f"Model R Squared score: {score}")

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = model.coef_
intercept = model.intercept_

print(f"Coefficients: {coefficients}")
print(f"Intercept: {intercept}")

Predicted output: 504067.2125712681
Model R Squared score: 0.8552472077276095
Coefficients: [ 866.14641337 1032.69506649]
Intercept: -409391.47958340764


#### Explanation of R^2 Score:
The R^2 Score, or coefficeint of determination, is a value that represents how well the model explains the variance in the target variable. The score ranges from 1 to 0, where a score of 1 indicates that the model perfectly predicts the target variable, ad a score of 0 indicates the model doesn't explain any of the variance in the data. So, the score of my model, 0.855, indicates that my model is explaining about 85.5% of the variance in the target variable "Size nm^3". This score is high, suggesting that the model is performing relatively well, but there is still some room for improvement.

#### Equation for h(x): 

$$
h(\text{Temperature °C}, \text{Mols KCL}) = -409391.48 + 866.15 \cdot \text{Temperature °C} + 1032.70 \cdot \text{Mols KCL}
$$


## Part 4. Use Cross Validation

In [14]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
cv_scores = cross_val_score(model, X, y, cv = 5)

# Report on their finding and their significance
print(f"Cross-validation scores: {cv_scores}")
print(f"Mean Cv Score: {cv_scores.mean()}")
print(f"Std. Deviation of CV Scores: {cv_scores.std()}")

Cross-validation scores: [0.83918826 0.87051239 0.85871066 0.87202623 0.84364641]
Mean Cv Score: 0.8568167899144437
Std. Deviation of CV Scores: 0.01346630737209602


#### Cross-Validation Findings: 
The cross-validation scores indicate consistent performance across the five data splits, ranging from approximately 0.839 to 0.872. With a mean cross-validation score of 0.857, the model explains about 85.7% of the variance in the target variable, demonstrating solid overall performance. Additionally, a standard deviation of 0.0013 reveals minimal variability across the folds, which underscores the model's reliable performance. This low standard deviation further suggests that the model is unlikely to be sensitive to specific subsets of the dataset. Overall, these metrics imply that the model is effective and generalizes well to unseen data.

## Part 5. Using Polynomial Regression

In [19]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(degree = 2)
X_train_poly = poly.fit_transform(X_train)
X_test_poly = poly.transform(X_test)

# Train Model
poly_model = LinearRegression()
poly_model.fit(X_train_poly, y_train)

# Sample datapoint
sample_datapoint_poly = poly.transform(sample_datapoint)

predicted_value_poly = poly_model.predict(sample_datapoint_poly)
print(f"Predicted output: {predicted_value_poly[0]}")

score_poly = poly_model.score(X_test_poly, y_test)
print(f"R squared score: {score_poly}")

# Report on the metrics and output the resultant equation as you did in Part 3.
coefficients_poly = poly_model.coef_
intercept_poly = poly_model.intercept_

print(f"Coefficients: {coefficients_poly}")
print(f"Intercept: {intercept_poly}")



Predicted output: 477057.38678902347
R squared score: 1.0
Coefficients: [ 0.00000000e+00  1.20000000e+01 -1.27196606e-07  1.26450728e-11
  2.00000000e+00  2.85714287e-02]
Intercept: 2.0478793885558844e-05


#### Evaluation of Performance:
The polynomial regression model achieved an R² score of 1.0, indicating that it explains 100% of the variance in the target variable. While this suggests a perfect fit to the training data, it raises significant concerns about potential overfitting. Overfitting occurs when the model captures not only the underlying patterns but also the noise within the training dataset, potentially leading to poor generalization on unseen data. Thus, despite the high R² score, further evaluation on a separate test set is crucial to ensure the model's robustness and its ability to accurately predict outcomes in real-world applications.

### Equation for h(x):
$$
\begin{align*}
h(x) &= 2.0478793885558844 \times 10^{-5} \\
&+ 12 \cdot \text{Temperature} \\
&+ (-1.27196606 \times 10^{-7}) \cdot \text{Mols KCL} \\
&+ (1.26450728 \times 10^{-11}) \cdot \text{Temperature}^2 \\
&+ 2 \cdot (\text{Temperature} \cdot \text{Mols KCL}) \\
&+ 0.02857143 \cdot \text{Mols KCL}^2
\end{align*}
$$
